### PR TITLE
Merging PR #52 — Course 1 Exercises (Part 3, Done)

### PR DESCRIPTION
**Pull request by heng-yuwen**
# ✈  [**Done**] Merging PR #52 — Merging changes to Course 1 - Computer Vision [Exercises](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/tree/main/1-Computer-Vision/Exercises).

Includes API updates to [TensorFlow v2.11.0](https://github.com/tensorflow/tensorflow/releases/tag/v2.11.0) and [Waymo Open Dataset](https://github.com/waymo-research/waymo-open-dataset/commit/c9bf4d829bdf10f48a0dec2ef0ef2e5e05a8a0d). Also included are brief performance evaluations of the image classification task on Apple Silicon hardware. To view the diffs for `.ipynb` files, use a Jupyter notebook diff renderer, e.g., [ReviewNB](https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/).

  

**Please check if the PR fulfills these requirements**:
- [x] The commit message follows repository guidelines;
- [x] Tests for the changes have been added (for bug fixes / features);
- [x] Docs have been added / updated (for bug fixes / features).

## :label: Type of change(s)

- [x] :sparkles: New feature (non-breaking change which adds functionality);
   - [x] :arrow_up: Upgraded dependency—TensorFlow v2.8 to v2.11.0;
   - [x] :arrow_up: Upgraded dependency—`waymo-open-dataset-tf-2-6-0` to `waymo-open-dataset-tf-2-11-0`; 
- [x] :memo: This change requires a documentation update;
   - [x] :pushpin: Pinned dependencies for Waymo Open Dataset API:
      * Fetching latest available commit #[c9bf4d8](https://github.com/waymo-research/waymo-open-dataset/commit/c9bf4d829bdf10f48a0dec2ef0ef2e5e05a8a0db), installing `waymo-open-dataset-tf-2-11-0==1.6.0` using `pip`;
   - [x] :pushpin: Pinned dependencies for TF2 Object Detection API:
      * The dependency version conflicts have been solved with the addition of the [*`setup.py`*](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/commits/b31bc50e146211cbd55d523463e78b5cf4745210#diff-6f10d92b4993c2cdd818fb4f6f7f414c5c38313e29d1197b56bc28abcbcdcde8) file — NOTE: this is pulled directly from the official [TensorFlow 2 Object Detection API](https://github.com/tensorflow/models/blob/master/research/object_detection/packages/tf2/setup.py) repository; 
   - [x] :arrow_down: Downgraded dependency—`cython<3.0.0` (WORKAROUND: intentionally downgrading to resolve TF Object Detection API installation issues).

 

## 🧪 Testing and verification
In [`../2022-08-03-Creating-TF-Records.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-1-3-Creating-TF-Records/2022-08-03-Creating-TF-Records.ipynb):
- [x]  :white_check_mark: `model_builder_tf2_test.py`: the TF2 `object_detection` API passed the official build test after upgrade to TensorFlow V2.11.0:
   * Completed `23/24` tests (1 skipped)—passes with `OK`.

In [`../2022-09-05-Image-Classification-Feed-Forward-Neural-Networks.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/1c667a019600442e15d8d2aa3c9c3a7ce4c8f71f/1-Computer-Vision/Exercises/1-3-3-Image-Classification-FNNs/2022-09-05-Image-Classification-Feed-Forward-Neural-Networks.ipynb):
- [x] :white_check_mark: Training metrics (loss, accuracy) on the Image classification with FFNNs notebook are similar after upgrade to TF v2.11.0 and after modification of learning rate and LR schedule.

## 💬 More info

The following notebook files have been added / modified:
* [`../2022-08-03-Creating-TF-Records.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-1-3-Creating-TF-Records/2022-08-03-Creating-TF-Records.ipynb):
   * `waymo-open-dataset-tf-2-11-0==1.6.0`:  using the latest `waymo-open-dataset` version to support upgrade to TensorFlow V2.11.0.
   * `ENV_COLAB = False`: by default, this notebook is run using personal device rather than Colab.
   * `google.auth`: allow to login to Google using  google cloud sdk.
   * `object_detection API`: use commit 15ab048c9f from Nov 2022, just the same day as the release of [models v2.11.0](https://github.com/tensorflow/models/releases/tag/v2.11.0).
   * `apt install -y protobuf-compiler`: pip install failed on Ubuntu 20.04, so use apt to install protobuf-compiler instead.
   * `cython<3.0.0`: cython 3.0.0 ruined the notebook, downgrade so that the object_detection API can be installed.

* [`../2022-09-05-Image-Classification-Feed-Forward-Neural-Networks.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-3-3-Image-Classification-FNNs/2022-09-05-Image-Classification-Feed-Forward-Neural-Networks.ipynb):
  * `ENV_COLAB = False`: by default, this notebook is run using personal device rather than Colab.
  * `tf.keras.metrics.SparseCategoricalAccuracy(name="accuracy")`: use proper accuracy class when using a self-defined optimiser and scheduler;
  * `decay = True`: using the learning rate schedule with a decay rate of `0.96`;
  * `model_params.update({'learning_rate': 1e-3})`: decreasing learning rate for the `'adam'` optimiser.

* [`../setup.py`](https://github.com/heng-yuwen/ND0013-Self-Driving-Car-Engineer/blob/pr49-course-1-revision-1-heng-yuwen/1-Computer-Vision/Exercises/1-1-3-Creating-TF-Records/setup.py)
  * `tensorflow==2.11.0`: specify tensorflow version while installing Object Detection API. The file *setup.py* is copied from [official](https://github.com/tensorflow/models/blob/15ab048c9fb3864a5604a300c7148b4b830c3c9b/research/object_detection/packages/tf2/setup.py) git repo;
  * `numpy==1.21.5`, `pyarrow==10.0.0`: add version requirements specified by `waymo-open-dataset`.

### CODE REVIEW

File: 1-Computer-Vision/Exercises/1-3-1-Logistic-Regression/2022-08-27-Logistic-Regression.ipynb [hunk 1]
```diff
@@ -80,7 +80,7 @@
   },
```
Comment (jonathanloganmoran) [inline 1-Computer-Vision/Exercises/1-3-1-Logistic-Regression/2022-08-27-Logistic-Regression.ipynb:80]: <p>I see that this is quite a tremendous performance hit; hope you are able to regain access to your GPU server soon.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1299277533'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='58'/>

File: 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb [hunk 2]
```diff
@@ -46,7 +46,7 @@
   },
```
Comment (jonathanloganmoran) [inline 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>Re:</p><blockquote>Looks like mean validation accuracy dropped slightly in your experiment run.</blockquote><p> </p><p>Looks like with the Apple Silicon architecture you have "enough" cache to fit the batch in memory (running M2 Max, correct?). Not sure your training batch size would be automatically reduced because of this...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1299277537'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (jonathanloganmoran) [inline 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>Re:</p><blockquote>Looks like mean validation accuracy dropped slightly in your experiment run.</blockquote><p> </p><p>From this sample, I see no obvious changes to the GTSRB training images. Therefore, I wouldn't suspect any data corruption or issues with the image generator just yet. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1299277539'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

Comment (jonathanloganmoran) [inline 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>Looks like mean validation accuracy dropped slightly in your experiment run.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1299277540'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='92'/>

**Comment by jonathanloganmoran** [comment]
This is wonderful work—thank you, @heng-yuwen. My review is complete, please see the comments left above on [ReviewNB](https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/). Also, apologies for the mixed posting here; looks like this bot integration isn't perfect. 

Happy to see you get quickly up-to-speed with the TF v2.11.0 upgrade. Congratulations for discovering which dependency versions work with this project. This will save much headache for the package manager and for other users in the future.

Will be glad to merge #52 into [`jonathanloganmoran:pr49-course-1-revision-1-heng-yuwen`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/tree/pr49-course-1-revision-1-heng-yuwen) once your revisions / responses here are complete.

**Comment by heng-yuwen** [comment]
<p>Comment added</p>

---

 <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1685128297'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by heng-yuwen** [comment]
<p>I haven't experienced any data loss in my experiments so far, but we can have a look at crcmod later.</p>

---

 <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1685128299'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by heng-yuwen** [comment]
<p>I tested the latest one, but it does not work well with tf 2.11.0. I don't have time to figure out why, but this commit is surely designed for tf 2.11.0. So, let's stay with this one for now.</p>

---

 <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1685128302'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by heng-yuwen** [comment]
<p>I solved this with a better way, by providing a <strong><em>setup.py </em></strong>which explicitly asks for tensorflow==2.11.0</p>

---

 <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1685128306'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb (hunk 2 above)
Comment (heng-yuwen) [reply 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>That happens due to random initialisation of model parameters. Rerun it several times and the curves are totally different.</p>

**Comment by jonathanloganmoran** [comment]
<p>Thanks for investigating. Marking this as resolved.</p>

---

 <a href='https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/52/discussion/#comment-1685128302'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb (hunk 2 above)
Comment (jonathanloganmoran) [reply 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>Would setting a fixed global seed help address this? Assuming we can call <code>tf.random.set_seed</code> prior to initialising the model weight vector <code>W</code>, as shown below:</p><p> </p><h3><code># Setting a fixed global seed prior to initialising model weight vector</code></h3><h3><code>tf.random.set_seed(seed=42)</code></h3><p><code>W = tf.Variable(tf.random.normal(</code></p><p><code>        shape=(</code></p><p><code>            model_params['num_inputs'],&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           model_params['num_outputs']</code></p><p><code>        ),&nbsp;</code></p><p><code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;mean=0,&nbsp;</code></p><p><code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;stddev=0.01</code></p><p><code>))</code></p>

Comment (heng-yuwen) [reply 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>Rerun everything using GPU now. It is weird that the training speed of Nvidia 3090 is not as fast as Apple M2. It is aware that there is about 20% performance decay using docker. The good news is that the final performance of the trained model seems fine now.</p>

Comment (heng-yuwen) [reply 1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb:46]: <p>Yes, setting a global seed is a choice. I am not familar with TensorFlow, but with PyTorch, you need to fix the seed of numpy as well, and enable the deterministic mode as well. </p><p> </p><p>Also, fixing the seed is usually used when comparing the performance of multiple models. Usually seed search should be performed as well to find the best model. Here I think it is fine to train the model without fixed random seed.</p>

**Comment by heng-yuwen** [comment]
Hi @jonathanloganmoran , could you please kindly review my responses and merge this PR? I will have a rest for the next few days, and continue reviewing the notebooks soon.

**Comment by jonathanloganmoran** [comment]
> Hi @jonathanloganmoran , could you please kindly review my responses and merge this PR? I will have a rest for the next few days, and continue reviewing the notebooks soon.

Yes, absolutely. Will get back to you with my review in the next few hours. Thank you and please enjoy your well-deserved rest.
